### DISCUSSION TITLE
Interfaces composed from delegates

### DISCUSSION THREAD

**Discussion by jochenz** [dotnet/csharplang#5660] [Language Ideas]
A **delegate** is a type which represents a method signature.

An **interface** is type defined by a set of method signatures.

So, shouldn't we be able to compose interfaces from delegates like this?

``` cs
delegate double Add(double x, double y);
delegate double Subtract(double x, double y);
delegate double Multiply(double x, double y);
delegate double Divide(double x, double y);

interface ICalculator {
    Add; 
    Subtract;
    Multiply;
    Divide;
}
```

This would allow code to be very specific about what it needs (and to compose in a 'functional' way). 

It is clear that calculating an average does not require multiplying nor subtracting, so why would we inject an entire `ICalculator` in the `CalculateAverage` method?

``` cs
double CalculateAverage(IEnumerable<double> source, Add add, Divide divide) {
    // implementation here
}
```

Of course, this is an oversimplified example, but you can see the link with functional composition.

The entire idea can also work the other way around. Suppose you have an interface

``` cs
interface ICustomerRepo {
    Task<Customer?> GetById(CustomerId id);
    Task<Customer[]> Search(string term);
    Task SaveChanges();
}
```

Then why would we not allow C# to treat the individual interface members as delegates?

``` cs
async Task ActivateCustomer(
    CustomerId id, 
    ICustomerRepo.GetById getCustomer, 
    ICustomerRepo.SaveChanges saveChanges) {
        // overly simplified implementation without null case
        var customer = await getCustomer();
        customer.Activate();
        await saveChanges();        
}
```

I'm not saying these are the best examples and the second strategy would need to deal with overloads (not supported on delegates), but that can be overcome through more extensive syntax (add arg list of choice).

I'm primarely interested in the first case:

- define a delegate
- use the delegate in an interface member declaration
- test the code without the need to implement the interface

What do you think?

**Comment by HaloFour** [comment]
How would that work for interfaces with overloaded members?

**Reply by jochenz** [reply]
Simply by allowing more explicit choice through a more verbose syntax:

``` cs
interface ICalculator {
    int Add(int x, int y);
    double Add(int x, double y);
}

void MyMethod(Shipment x, ICalculator.Add(int, double) add) {
}
```

But, as written before, don't focus on using interface members as delegates, it would already be nice to compose interfaces from delegates.

**Reply by HaloFour** [reply]
What would the name of the delegate type be?  What creates that delegate?  What about any existing interfaces where that delegate has not been created?

More importantly, what does this enable?  What kind of problems is this trying to solve?  Why would you _want_ to treat an interface member as a delegate, and why doesn't the existing delegate functionality suffice?  What makes interface members different than any other members?

**Comment by CyrusNajmabadi** [comment]
> So, shouldn't we be able to compose interfaces from delegates like this?

You can basically do this already:

```c#
public interface ICalculator {
    Add Add { get; }
    Subtract Subtract { get; }
    Multiply Multiply { get; }
    Divide Divide { get; }
}
```

**Reply by jochenz** [reply]
But that's not really the same, is it?
It looks the same for the consumer, but it is a convoluted implementation (properties returning delegates), instead of normal virtual methods behind the scene.
It requires a different & unusual interface implementation and would maybe also require special compiler treatment to be performant (don't have the background to judge that).

**Reply by CyrusNajmabadi** [reply]
> but it is a convoluted implementation (properties returning delegates),

I mean... i thought that's what you wanted :)  you were the one that put the delegates in the interface :)
